pkg/core/native/name_service.go

```diff
@@ -0,0 +1,727 @@
+		manifest.NewParameter("type", smartcontract.IntegerType))
+	md = newMethodAndPrice(n.deleteRecord, 1000000, callflag.WriteStates)
+	n.AddMethod(md, desc)
+
+	desc = newDescriptor("resolve", smartcontract.StringType,
+		manifest.NewParameter("name", smartcontract.StringType),
+		manifest.NewParameter("type", smartcontract.IntegerType))
+	md = newMethodAndPrice(n.resolve, 3000000, callflag.ReadStates)
+	n.AddMethod(md, desc)
+
+	return n
+}
+
+// Metadata implements interop.Contract interface.
+func (n *NameService) Metadata() *interop.ContractMD {
+	return &n.ContractMD
+}
+
+// Initialize implements interop.Contract interface.
+func (n *NameService) Initialize(ic *interop.Context) error {
+	si := &state.StorageItem{Value: bigint.ToBytes(big.NewInt(DefaultDomainPrice))}
+	return ic.DAO.PutStorageItem(n.ContractID, []byte{prefixDomainPrice}, si)
+}
+
+// OnPersist implements interop.Contract interface.
+func (n *NameService) OnPersist(ic *interop.Context) error {
+	return nil
+}
+
+// PostPersist implements interop.Contract interface.
+func (n *NameService) PostPersist(ic *interop.Context) error {
+	return nil
+}
+
+func (n *NameService) addRoot(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	root := toString(args[0])
+	if !rootRegex.Match([]byte(root)) {
+		panic("invalid root")
+	}
+
+	n.checkCommittee(ic)
+	roots, _ := n.getRootsInternal(ic.DAO)
+	if !roots.add(root) {
+		panic("name already exists")
+	}
+
+	err := putSerializableToDAO(n.ContractID, ic.DAO, []byte{prefixRoots}, &roots)
+	if err != nil {
+		panic(err)
+	}
+	return stackitem.Null{}
+}
+
+var maxPrice = big.NewInt(10000_00000000)
+
+func (n *NameService) setPrice(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	price := toBigInt(args[0])
+	if price.Sign() <= 0 || price.Cmp(maxPrice) >= 0 {
+		panic("invalid price")
+	}
+
+	n.checkCommittee(ic)
+	si := &state.StorageItem{Value: bigint.ToBytes(price)}
+	err := ic.DAO.PutStorageItem(n.ContractID, []byte{prefixDomainPrice}, si)
+	if err != nil {
+		panic(err)
+	}
+	return stackitem.Null{}
+}
+
+func (n *NameService) getPrice(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.NewBigInteger(n.getPriceInternal(ic.DAO))
+}
+
+func (n *NameService) getPriceInternal(d dao.DAO) *big.Int {
+	si := d.GetStorageItem(n.ContractID, []byte{prefixDomainPrice})
+	return bigint.FromBytes(si.Value)
+}
+
+func (n *NameService) parseName(item stackitem.Item) (string, []string, []byte) {
+	name := toName(item)
+	names := strings.Split(name, ".")
+	if len(names) != 2 {
+		panic("invalid name")
+	}
+	return name, names, n.getTokenKey([]byte(name))
+}
+
+func (n *NameService) isAvailable(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	_, names, key := n.parseName(args[0])
+	if ic.DAO.GetStorageItem(n.ContractID, key) != nil {
+		return stackitem.NewBool(false)
+	}
+
+	roots, ok := n.getRootsInternal(ic.DAO)
+	if !ok {
+		panic("missing roots")
+	}
+
+	_, ok = roots.index(names[1])
```

Comment:
`ok` means that the domain is already registered and is _not_ available.

Reply:
What we are doing here is checking if _domain_ was previously registered (and thus name containing this domain is available).
In C# code we actually panic in case of `false`, fixed this.

Reply:
Also added test to cover this possibility.